InformationScripting/src/queries/TagQuery.cpp

```diff
@@ -0,0 +1,124 @@
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#include "TagQuery.h"
+
+#include "QueryRegistry.h"
+#include "../nodes/TagExtension.h"
+
+#include "ModelBase/src/nodes/composite/CompositeNode.h"
+
+namespace InformationScripting {
+
+const QStringList TagQuery::TAGTYPE_ARGUMENT_NAMES{"t", "type"};
+const QStringList TagQuery::NAME_ARGUMENT_NAMES{"n", "name"};
+
+QList<TupleSet> TagQuery::execute(QList<TupleSet> input)
+{
+	return exec_(this, input);
+}
+
+void TagQuery::registerDefaultQueries()
+{
+	QueryRegistry::instance().registerQueryConstructor("tags", [](Model::Node* target, QStringList args) {
+		return new TagQuery(&TagQuery::queryTags, target, args);
+	});
+	QueryRegistry::instance().registerQueryConstructor("addTags", [](Model::Node* target, QStringList args) {
+		return new TagQuery(&TagQuery::addTags, target, args);
+	});
+}
+
+TagQuery::TagQuery(ExecuteFunction<TagQuery> exec, Model::Node* target, QStringList args)
+	: ScopedArgumentQuery{target, {
+			{TAGTYPE_ARGUMENT_NAMES, "Tag Type argument", TAGTYPE_ARGUMENT_NAMES[1]},
+			{NAME_ARGUMENT_NAMES, "Name of a symbol", NAME_ARGUMENT_NAMES[1]},
+		}, QStringList("TagQuery") + args}, exec_{exec}
+{}
+
+QList<TupleSet> TagQuery::queryTags(QList<TupleSet> input)
+{
+	QList<TupleSet> result;
+	if (scope() == Scope::Local)
+	{
+
+	}
+	else if (scope() == Scope::Global)
+	{
+
+	}
+	else if (scope() == Scope::Input)
+	{
+		Q_ASSERT(input.size() > 0);
+		TupleSet tupleSet = input.takeFirst();
+		auto astTuples = tupleSet.tuples("ast");
+		for (auto tuple : astTuples)
+		{
+			Model::Node* node = tuple["ast"];
+			if (auto astNode = DCast<Model::CompositeNode>(node))
+			{
+				auto tagExtension = astNode->extension<TagExtension>();
+				if (auto tagNode = tagExtension->tag())
+					qDebug() << tagNode->name();
+			}
+		}
+		result << tupleSet;
+	}
+	return result;
+}
+
+QList<TupleSet> TagQuery::addTags(QList<TupleSet> input)
+{
+	QList<TupleSet> result;
+	if (scope() == Scope::Local)
+	{
+
+	}
+	else if (scope() == Scope::Global)
+	{
+
+	}
+	else if (scope() == Scope::Input)
+	{
+		Q_ASSERT(input.size() > 0);
+		TupleSet tupleSet = input.takeFirst();
+		auto astTuples = tupleSet.tuples("ast");
+		for (auto tuple : astTuples)
+		{
+			Model::Node* node = tuple["ast"];
+			if (auto astNode = DCast<Model::CompositeNode>(node))
+			{
+				auto tagExtension = astNode->extension<TagExtension>();
+				astNode->beginModification("addTag");
```

Comment:
Brilliant. You've just implemented your first query that does modifications. I like it. I like how natural it feels to add tags this way. There might be other ways, but this one feels very natural, especially for adding tags on mass.